### ISSUE TITLE
Database backend abstractions for records

### ISSUE THREAD

**Issue by hrbonz** [gams/openkongqi#43]
Similar to #42 for records saving.

**Comment by eddowh** [comment]
I gotta be honest, for the first time I am 120% clueless and don't know how to move forward with this. I added a `records` entry in `global_settings['DATABASES']` like this:

```python
DATABASES = {
    'status': {
        [...],
    },
    'records': {
        'ENGINE': '',  # sqlite / pgsql / influxdb
        [...],
    },
}
```

and I similarly created a folder like:

```
openkongqi/records
├── base.py
└── __init__.py
```

where `base.py` has a class `BaseRecordsWrapper`, which has the following methods:

- `__init__`
- `create_cnx` (requires to be overwritten by children)
- `get_records` (requires to be overwritten by children)
- `set_records` (requires to be overwritten by children)

I don't know many things, especially how to start is how to initiate a connection. From what I can see in `openkongqi.db`, a database engine is initiated with `sqlalchemy.create_engine('sqlite:///openkongqi.db')`...? 

I think I need a lot of assistance here.

**Comment by hrbonz** [comment]
Create a generic sqlalchemy wrapper that inherits from your `BaseRecordsWrapper` to handle relational databases. When you have this you can create a `SQLiteWrapper` that inherits from `SQLAlchemyWrapper` and just configures the engine.

You must create a `Record` model that would probably have the following fields:
* uuid
* timestamp (**in UTC!!!!1**)
* measurement: name of the measure could be "pm25" or "pm10", etc
* value: the value for this measurement

**Comment by eddowh** [comment]
How will this `Record` model differ from the already existing `Datapoint` model? Shouldn't we able to reuse this?

```python
class Datapoint(Base):
    __tablename__ = 'datapoint'

    ts = Column(DateTime, primary_key=True)
    source = Column(String(250), nullable=False, primary_key=True)
    station = Column(String(250), nullable=False, primary_key=True)
    key = Column(String(250), nullable=False, primary_key=True)
    value = Column(Float(), nullable=True)
```

**Comment by hrbonz** [comment]
Here's a bit of pseudo code that should get you started.

```python
class SQLAlchemyRecordsWrapper(object): 
    def __init__(self):
        pass

    def create_cnx(self, settings):
        """
        configuration looks like:
        DATABASES = {
            'records': {
                'ENGINE': "openkongqi.records.pgsql",
                'HOST': xxx,
                'PORT': xxx,
                'USER': xxx,
                'PWD': xxx,
                'NAME': xxx 
            }
        }
        """
        engine = create_engine(settings['BACKEND'])
        db_session = scoped_session(sessionmaker(autocommit=False,
                                                 autoflush=False,
                                                 bind=engine))
        Base = declarative_base()
        Base.query = db_session.query_property()

    def write_records(self, records):
        for record in records:
            r = Record(record)
            r.save()

    def get_records(self, start, end, filters=None):
        """Returns a list of Records"""
        if filters is None:                      
            q = "SELECT * from table WHERE ts >= start AND ts <= end;"
        else:
            q = "SELECT {} from table WHERE ts >= start AND ts <= end;".format(filters.join(", "))
        # FIXME 
        return Record.query(q)
        

class Record(Base):
    __tablename__ = 'records'
        
    ts = Column(DateTime, primary_key=True)
    uuid = Column(String(250), nullable=False, primary_key=True)
    key = Column(String(250), nullable=False, primary_key=True)
    value = Column(Float(), nullable=True)
```

**Comment by eddowh** [comment]
I don't think the model `Record` should be in the `openkongqi.sqlalchemy_wrapper` module. From what I can see with what you did in `openkongqi.source.base`, you:

1. create database `engine`, which requires:
    - `sqlalchemy.create_engine`
2. create `db_session`, which requires:
    - `sqlalchemy.orm.scoped_session`
    - `sqlalchemy.orm.sessionmaker`
3. create `Record` model by creating `Base` class, which requires:
    - `sqlalchemy.ext.declarative.declarative_base`
4. use `db_session` for querying and writing records

The problem right now is that if we create `engine` and `db_session` from within the `SQLAlchemyRecordsWrapper` class, the `Record` model has no way of accessing it.

However, the good news is that we just need to create a valid DSN (data source name) with specific modules (i.e. `openkongqi.records.sqlite_db`) in the `openkongqi.conf` module, then create a `recs_db_session`.

Full details are in the gist here: <https://gist.github.com/eddowh/6d6454f395a16c3b4a5c>

NOTE: GitHub Gist doesn't support **/** in filenames, so files from subdirectories are shown like:

`openkongqi/records/base.py => records.base.py`

**Comment by hrbonz** [comment]
Few comments here:
* just realised `.status.base.BaseStatusWrapper` should inherit from `object` (already changed)
* make both status and records base classes accept `*args` and `**kwargs` in their `__init__`
* move `create_status_db_instance` from `.conf` to `.status.base` and rename to `create_statusdb()`
* for records, write `create_recsdb()` and hide the session creation in there, return a wrapper instance
* Use [sqlalchemy.engine.url.URL](http://docs.sqlalchemy.org/en/rel_1_0/core/engines.html#sqlalchemy.engine.url.URL) to create your dsn
* add `__repr__()` method to the Record model
* Do not change the `Base` class query engine and use `session` directly to query the database

**Comment by hrbonz** [comment]
Closed by 609687a740509b65ee5b1264b81c652590b5dea4
